Support implicit 'to' version as latest

rodrigues opened the pull request:
Hi there 👋

First, congrats for the feature, it's really cool!

I've been using the CLI diff for a bit, and the thing I usually want is:

**Show me the diff of the package from current installed version to latest _stable_ available.**

The currently installed version is known locally to the user (latest isn't), so it would be nice if the diff tool could accept a route with an implicit latest, so we can have a local command that opens quickly the browser.

What are your thoughts? 🙂

This code seems to do it nice, please tell me if you'd like any changes 👍

Thanks!

ericmj commented on lib/diff_web/controllers/page_controller.ex:
```diff
@@ -36,14 +36,29 @@ defmodule DiffWeb.PageController do
+      version =
+        versions
+        |> Enum.map(&Version.parse!/1)
+        |> Enum.max()
```
We need to use the new `module.compare/2` sorting instead of term based sorting.

```suggestion
        |> Enum.max(Version)
```

rodrigues replied:
Nice, I've changed in here https://github.com/hexpm/diff/pull/45/commits/b9071e34c235a14886d5f9aa03c0bf3ab8a63e0f 👍

ericmj commented on lib/diff_web/controllers/page_controller.ex:
```diff
@@ -36,14 +36,29 @@ defmodule DiffWeb.PageController do
+  defp parse_versions(package, versions) do
+    with [from, to] <- versions |> String.split("..") |> Enum.map(&String.trim/1),
+         {:ok, from} <- parse_version(package, :from, from),
+         {:ok, to} <- parse_version(package, :to, to) do
```
What happens `:from` is already the latest version?

rodrigues replied:
Good point! After this other commit https://github.com/hexpm/diff/pull/45/commits/b9071e34c235a14886d5f9aa03c0bf3ab8a63e0f , it is 400'ing when from is empty (e.g. "/diff/package/%20.."), and also when the versions are the same. Would you want a 400 in this last case?

rodrigues commented:
Thanks for the feedback, @ericmj and @wojtekmach, I should find time later tonight / this weekend to try addressing your points 👍

Of course, feel free to attack it too if you want to.

rodrigues commented on lib/diff_web/controllers/page_controller.ex:
```diff
@@ -36,14 +36,29 @@ defmodule DiffWeb.PageController do
+  defp parse_version(_package, :from, ""), do: :error
+
+  defp parse_version(package, :to, "") do
+    with {:ok, versions} <- Diff.Package.Store.get_versions(package) do
```
Good points, I've changed latest to discard rc's (indeed that's what was happening), and managed to move non-pure calls to `do_diff`, you think now it's OK? https://github.com/hexpm/diff/pull/45/commits/b9071e34c235a14886d5f9aa03c0bf3ab8a63e0f

rodrigues commented:
Hey! Take your time, of course. Tell me later if you want me to squash these two commits 👍

wojtekmach commented on test/diff_web/controllers/page_controller_test.exs:
```diff
@@ -1,8 +1,29 @@
     assert html_response(conn, 200) =~ "<!DOCTYPE html>"
   end
+
+  describe "GET /diff/:package/:versions" do
```
thank you for adding the test!

wojtekmach commented on test/diff_web/controllers/page_controller_test.exs:
```diff
@@ -1,8 +1,29 @@
+      expect(Diff.Package.StoreMock, :get_versions, fn "phoenix" -> {:ok, versions} end)
+      expect(Diff.StorageMock, :get, fn "phoenix", "1.4.5", "1.4.9" -> {:ok, diff} end)
+
+      conn = get(conn, "diff/phoenix/1.4.5..")
```
actually, I think this URL is a bit awkward. My other ideas would be:

1. `diff/phoenix/1.4.5`
2. `diff/phoenix/1.4.5..latest`

I think I like option 1. the most. What do you all think?

rodrigues replied:
Hey! I'm fine with the url with the `..` hehe, but I understand the awkwardness.

Between 1 and 2 I prefer 2, since the idea is to be clear, 2 seems to me the clearest of them all.

rodrigues replied:
Will go that path (2), and squash all the things, in 3, 2, 1… 🙂

wojtekmach replied:
before making any changes let's wait for other opinions!

wojtekmach replied:
btw, no worries about squashing commits, we almost always squash when merging.

ericmj replied:
I think `diff/phoenix/1.4.5` looks best.

rodrigues replied:
thanks for the input! should we go with `diff/phoenix/1.4.5` then? 🙂

rodrigues replied:
https://github.com/hexpm/diff/pull/45/commits/b7127a5adf583af615abf18523dd4cde7af725f9 should do it 👍

good thing that `diff/phoenix/1.4.5..` incidentally also work 😄

wojtekmach replied:
I think `diff/phoenix/1.4.5..` should be a 400 bad request but I don't have a strong opinion about it.

ericmj commented:
This is great, thank you!
